File: 1032.py
#!/usr/bin/env python

# timeout for the last case
# whether or not using dict 

if __name__ == '__main__':
    li = raw_input().split()
    addr_start1 = li[0]
    addr_start2 = li[1]
    node_count = int(li[2])
    
    if addr_start1 == addr_start2:
        print addr_start1
        import sys
        sys.exit(0)

    node_map = {}
    for i in xrange(node_count):
        li = raw_input().split() 
        addr = li[0]
        node_map[addr] = li[2]

    mark_list = {}
    st = addr_start1
    pos = '0'
    while st != '-1':
        mark_list[st] = '1'
        st = node_map[st]

    st = addr_start2
    while st != '-1':
        if mark_list.get(st, '-1') == '1':
           pos = st
           break
        st = node_map[st]

    if pos == '0':
        print -1
    else:
        print pos

File: 1032.cpp
#include <iostream>
using namespace std;

int main()
{
    int addr_start1, addr_start2, addr, n, next;
    char ch;
    int node_map[100000];
    int mark_map[100000];
    
    cin >> addr_start1 >> addr_start2 >> n;
    
    for (int i = 0; i < n; i++)
    {
        cin >> addr >> ch >> next;
        node_map[addr] = next;
    }

    while (addr_start1 != -1)
    {
        mark_map[addr_start1] = 1;
        addr_start1 = node_map[addr_start1];
    }

    while (addr_start2 != -1)
    {
        if (mark_map[addr_start2] == 1)
        {
            cout.width(5);
            cout.fill('0');
            cout.setf(ios::right);
            cout << addr_start2 << endl;
            return 0;
        }
        addr_start2 = node_map[addr_start2];
    }
    cout << -1 << endl;

    return 0;
}
